### ISSUE TITLE
Update boost

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#3086]

**Comment by bangerth** [comment]
With GCC 6.2, I get a lot of new warnings from BOOST. The most common one is this one:
```
/home/bangerth/p/deal.II/1/dealii/bundled/boost-1.56.0/include/boost/function/function_base.hpp:318:13: warning: placement new constructing an object of type ‘boost::detail::function::functor_manager_common<std::_Bind<std::_Mem_fn<void (dealii::FEValuesBase<1, 1>::*)()>(std::reference_wrapper<dealii::FEValuesBase<1, 1> >)> >::functor_type {aka std::_Bind<std::_Mem_fn<void (dealii::FEValuesBase<1, 1>::*)()>(std::reference_wrapper<dealii::FEValuesBase<1, 1> >)>}’ and size ‘24’ in a region of type ‘char’ and size ‘1’ [-Wplacement-new=]
             new (reinterpret_cast<void*>(reinterpret_cast<void*>(&out_buffer.data))) functor_type(*in_functor);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
which happens dozens of times. (We get the `boost::function` stuff through `boost::signal`, which doesn't appear to use `std::function` even if C++11 or later is used.)

There are a couple of other issues as well. I see two options:
* Specific `-Wno-placement-new` to avoid this specific error
* Update BOOST to something more modern.

**Comment by drwells** [comment]
Yeah, this makes building with the bundled boost really annoying with GCC6. Unfortunately it looks like this was just fixed in July (boostorg/function#9) so boost will complain about this until 1.62 comes out (1.61 was out in May).

It would be nice to disable `-Wno-placement-new` only if we are using a user compiled copy of boost (so installed somewhere besides `/home/` or `/Users`/).

**Comment by bangerth** [comment]
Writing patch to disable the warning as we speak.

**Comment by bangerth** [comment]
I could apply that patch to the version of boost we bundle, which would allow us to enable the warning. But in the end, I also don't think that that's a warning that's going to trigger very frequently -- not in our code base, and likely not in our users' code bases either.

**Comment by drwells** [comment]
One thing that has come up a few times recently (#3138 and an improvement on what I proposed in #3263, in addition to a few other places in the library already) is the need for a class that contains a fixed size array (so memory allocations won't usually be needed) but can switch to an arbitrary size array in the free store if needed. The `boost::containers::small_vector` class does exactly this and is available in 1.58 (or newer).

I don't mind putting together a patch to update boost. Since we can't use 1.58 (broken serialization) does anyone have an objection with bundling the newest version of boost (1.62) and bumping the requirement (so that we can use `boost::containers::small_vector`) to 1.59?

**Comment by Rombur** [comment]
@drwells How is `small_vector` different that simply reserving the required size?

**Comment by bangerth** [comment]
@drwells -- that would be fantastic if you wanted to upgrade boost. I'm going to point out the difficulty in this step: boost is huge, and we have traditionally only bundled that part that we actually need. In the past, what I've done is to unpack the most recent boost, copy the whole tree to a new location, and then delete all files and subdirectories that are not also part of the currently bundled version. Something along the lines of
```
  cd new-boost
  for i in `find .` ; do
    if test -f $bundled-boost-dir/$i -o -d $bundled-boost-dir/$i ; then
      : ;      # currently bundled boost also has this file; do nothing
    else
       rm -r $i       # may yield errors because you may have already deleted $i through some higher-up directory
    fi
  done
```
Then copy into your deal.II tree and try to compile in a bunch of different configurations. You'll find that there are a bunch of dependencies that are now needed and that you accidentally deleted. Add these to the tree with the new bundled version by hand until you can compile successfully.

Lastly, to minimize git repo churn, the way to do the copy-over step is best done by doing a `git mv bundled/boost-1.56.0 bundled/boost-1.62.0` and then copy the new files over the old ones. Git then only has to store the deltas, rather than removing all files from the repository, and re-adding them in a separate commit. (I think I've read somewhere that git can be quite clever about this, but helping it keep the repository smallish -- in particular given the size of boost! -- surely can't hurt.)

**Comment by drwells** [comment]
@Rombur `small_vector` has a definition that looks like

```cpp
template <typename T, int N>
class small_vector
{
,,,
private:
  T data[N];
};
```

(or equivalent, there might be a `union` involved) so if the capacity is less than or equal to `N` the object does not need to allocate any memory in the free store (it just uses the buffer it owns inside itself). In contrast calling `std::vector<T>::reserve` allocates memory in the free store.

@bangerth I will give it a try; based on your warning this might take awhile.

**Comment by bangerth** [comment]
Yes, it's a bit of an involved process. Should be done with sufficient separation from the next release so we shake out the files we forgot to add but need on some platform or other.

**Comment by drwells** [comment]
I think that I got most of it working last night (see [feature branch](https://github.com/drwells/dealii/tree/update-boost-1.62)). Things seem OK with gcc-4.8 or newer.

Does anyone know what the oldest version of GCC that we officially support is? We cannot compile the library with GCC 4.1 for example.

**Comment by bangerth** [comment]
I have gcc 4.4 on my desktop. I think that was released in 2010, and it will be about as old as you will find today out in the wild.

**Comment by davydden** [comment]
i would be more than happy to require `4.8` and above :smile:

**Comment by kronbichler** [comment]
But 4.8 is too ambitious. I have a system with gcc 4.4 on it as well which is a nice baseline (but I know how to compile a newer compiler so this is no real problem for me).

**Comment by davydden** [comment]
> But 4.8 is too ambitious

agree.

**Comment by drwells** [comment]
Closed by #3279.
